### PR TITLE
[GROUND-26] db: Cache prepared statements

### PR DESCRIPTION
**Pull request by kevinji**
This PR is broken up into two commits: one part with just a file-level reformat, and the other with the real changes. I'm not quite happy with how connections are being closed at the moment; I'm hoping to use [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) to properly clean them up.

### CODE REVIEW

**Comment by shaneknapp** [comment]
ok to test

**Comment by vsreekanti** [comment]
Haven't looked at the code yet. I'll take a look on Tuesday.

From the CI test output, it looks like a lot (all?) of the Cassandra tests are failing.

File: ground-core/src/main/java/edu/berkeley/ground/api/models/EdgeFactory.java [hunk 1]
```diff
@@ -14,18 +14,17 @@
...
-import edu.berkeley.ground.db.DBClient.GroundDBConnection;
-import edu.berkeley.ground.exceptions.GroundException;
+import edu.berkeley.ground.exceptions.GroundDBException;
 
 import java.util.List;
 import java.util.Map;
 
 public abstract class EdgeFactory {
-  public abstract Edge create(String name, Map<String, Tag> tags) throws GroundException;
+  public abstract Edge create(String name, Map<String, Tag> tags) throws GroundDBException;
```
Comment (vsreekanti) [inline ground-core/src/main/java/edu/berkeley/ground/api/models/EdgeFactory.java]: There's no need for this change. `GroundDBException` is a subclass of `GroundException`. DB exceptions exist primarily for internal purposes. No need to expose them to world.

File: ground-core/src/main/java/edu/berkeley/ground/api/versions/cassandra/CassandraItemFactory.java [hunk 2]
```diff
@@ -93,23 +94,23 @@ public void update(GroundDBConnection connectionPointer, long itemId, long child
     }
 
     for (long parentId : parentIds) {
-      if (!(parentId == 0) && !dag.checkItemInDag(parentId)) {
+      if (parentId != 0 && !dag.checkItemInDag(parentId)) {
```
Comment (vsreekanti) [inline ground-core/src/main/java/edu/berkeley/ground/api/versions/cassandra/CassandraItemFactory.java:97]: lol, good catch. No idea what I was thinking when I wrote that...

File: ground-core/src/main/java/edu/berkeley/ground/db/CassandraClient.java [hunk 3]
```diff
@@ -1,217 +1,202 @@
...
-
-      BoundStatement statement = new BoundStatement(session.prepare(prepString));
+      JGraphTUtils.addEdge(this.graph, nvFromId, nvToId);
+    }
 
-      int index = 0;
-      for (DbDataContainer container : insertValues) {
-        CassandraClient.setValue(statement, container.getValue(), container.getGroundType(), index);
+    String fields =
+        insertValues.stream().map(DbDataContainer::getField).collect(Collectors.joining(", "));
```
Comment (vsreekanti) [inline ground-core/src/main/java/edu/berkeley/ground/db/CassandraClient.java:96]: Do we know if there's overhead associated with converting `List`s to `Stream`s? I'm assuming they've optimized things enough for there not to be, but it'd be good to know for sure.

**Review by vsreekanti — CHANGES_REQUESTED** [review]
Things look great for the most part. The only thing I'm not sure about is the use of the DB exception in the factory APIs. Is there any particular reason for that?

File: ground-core/src/main/java/edu/berkeley/ground/api/models/EdgeFactory.java (hunk 1 above)
Comment (kevinji) [reply ground-core/src/main/java/edu/berkeley/ground/api/models/EdgeFactory.java]: Hmm fair enough, I was just trying to unify the use of the DB exceptions, as they were `GroundException` in one spot and `GroundDBException` in another spot (sometimes in the same file). Should I convert everything to `GroundException` instead?

File: ground-core/src/main/java/edu/berkeley/ground/db/CassandraClient.java (hunk 3 above)
Comment (kevinji) [reply ground-core/src/main/java/edu/berkeley/ground/db/CassandraClient.java:96]: There's probably some degree of overhead, but I don't think it should be significant. (Also, the code for this section hasn't changed from my previous PR; GitHub just doesn't show super great diffs.)

**Comment by vsreekanti** [comment]
Ignore this failure, sorry. In the middle of reconfiguring the PRB.

**Review by vsreekanti — APPROVED** [review]

**Pull request merged**